### PR TITLE
[spaceship] Fix 2FA to handle sms fallback for accounts with no trusted devices

### PR DESCRIPTION
**Pull request by jesiegel1**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
Issue: #16108 

If a user has 2FA enabled but is not logged into an Apple device with their Apple ID (no trusted devices), auth will default to SMS for verification. For an account with no trusted devices, if a user only has one trusted phone number associated with their account, an SMS verification code will automatically be sent to that number. If they have more than one trusted phone number, an SMS won’t be sent until it’s explicitly requested.

Trusted phone numbers have a separate endpoint, but Spaceship currently posts 2FA codes to the trusted devices endpoint for all cases (except when the default number env var is set). For accounts with no trusted devices, if a user with one trusted number enters the SMS code they automatically received, Spaceship exits with an Unauthorized Error. If the account has more than one trusted number, the user won’t have received a 2FA code at all and will have nothing to enter into the prompt.

The current workaround is to enter `sms` when prompted by Fastlane for the 2FA code which will force Fastlane to have the user select a phone number, request a new SMS code, and post the code to the correct endpoint.

### Description
Checks for `"noTrustedDevices"` field to determine SMS fallback, adds logic to handle the different possible cases, one number/multiple numbers, default number env var set/not set.

Also added new fixtures and a bunch of tests for handling the different verification paths, hitting the right endpoints, and just filling out coverage for the 2FA flow in general.

### Testing Steps
`bundle exec rspec ./spaceship/spec/two_step_or_factor_client_spec.rb` or run `Spaceship::Portal.login("<apple_id>")` from a rake task to manually test the behavior.

### Review Notes
* The Unauthorized Error people are seeing is actually a side effect - `Spaceship::TunesClient#handle_itc_response` doesn’t catch the invalid code error when hitting the wrong endpoint (same for wrong code but correct endpoint), so the rescue block on `line 143` doesn’t execute, `store_session` gets called, returns a 401, and exits with Unauthorized Error - I can open a separate PR for this, have an easy fix saved locally.
* I think the 2FA code can really use a refactor - the tight coupling between methods makes accounting for these new cases (and testing) a bit hacky, and can cause some things to break like the incorrect code recursive retry (which currently doesn't work bec of the broken invalid code error handling^) - I have a partial refactor saved locally.
* The `does not prompt user to choose number` test is brittle because the test will pass if `choose_phone_number` is ever removed or its name changes, but wasn’t sure how else to test it.
* Note that the `"noTrustedDevices"` field seems to only be present in the initial response for `GET /appleauth/auth` and not subsequent requests. Two other fields persist, `"mode"` and `"type"`, but seem to only be present once an sms has already been sent (e.g. initial response when one trusted number bec auto sent; after explicitly requesting an sms when two numbers). 
* I also have a bunch of Charles recordings for different cases which could be useful to save for the future, but would have to find a way to redact them properly.

### CODE REVIEW

**Comment by jesiegel1** [comment]
@googlebot I signed it!

File: spaceship/lib/spaceship/two_step_or_factor_client.rb
Comment (janpio) [inline spaceship/lib/spaceship/two_step_or_factor_client.rb]: maybe add a comment what this line means exactly?

File: spaceship/lib/spaceship/two_step_or_factor_client.rb
Comment (janpio) [inline spaceship/lib/spaceship/two_step_or_factor_client.rb]: maybe add a comment what this line means exactly?

Comment (janpio) [reply spaceship/lib/spaceship/two_step_or_factor_client.rb]: My guess: No device connected to account that could get pushed auth code

File: spaceship/lib/spaceship/two_step_or_factor_client.rb
Comment (janpio) [reply spaceship/lib/spaceship/two_step_or_factor_client.rb]: My guess: Exactly 1 phone number connected to account + ... (see above)

File: spaceship/lib/spaceship/two_step_or_factor_client.rb
Comment (janpio) [inline spaceship/lib/spaceship/two_step_or_factor_client.rb]: should this maybe be renamed and the `should_request_code` part be called explicitly before calling this method in the places where `should_request_code = true`? might make it clearer how the code "flows"

File: spaceship/lib/spaceship/two_step_or_factor_client.rb
Comment (janpio) [inline spaceship/lib/spaceship/two_step_or_factor_client.rb]: This whole `if/elsif/else` block feels ready for a refactoring now that we have 4 cases: ENV var, sms already sent, sms to be sent after choosing phone, and pushed code. Probably a future PR though to not make this one more complex than it already is?

**Review by janpio — COMMENTED** [review]
I like this! Seems like I totally ignored the SMS cases when I wrote this. Still don't understand how this now finally can fail and explode... Apple must have cleaned up something in how they handle their stuff.

File: spaceship/lib/spaceship/two_step_or_factor_client.rb
Comment (jesiegel1) [reply spaceship/lib/spaceship/two_step_or_factor_client.rb]: I gave the `should_request_code` a default value to keep the behavior of this method the same as prior to these changes unless `should_request_code` was explicitly provided, but given that this method isn't called anywhere else I agree we should make it more clear.

File: spaceship/lib/spaceship/two_step_or_factor_client.rb
Comment (jesiegel1) [reply spaceship/lib/spaceship/two_step_or_factor_client.rb]: 100%, and I have a partial refactor saved locally to better handle the expanded cases and loosen the coupling between methods (e.g. splitting up `request_two_factor_code_from_phone` which tightly couples requesting a new code with prompting the user for a 2fa code and building the payload). Definitely better to include as a separate PR though to minimize the potential for regressions.

File: spaceship/lib/spaceship/two_step_or_factor_client.rb
Comment (jesiegel1) [reply spaceship/lib/spaceship/two_step_or_factor_client.rb]: @janpio Sorry, I may have misunderstood, can you clarify if you mean explicitly passing a value for `should_request_code` rather than having a default value, or if you mean separating out the call to request a code into its own method?

Comment (janpio) [reply spaceship/lib/spaceship/two_step_or_factor_client.rb]: I meant renaming this method to something without "request" and have the request part in an explicit method call before. But doesn't matter too much.

Comment (jesiegel1) [reply spaceship/lib/spaceship/two_step_or_factor_client.rb]: I think this should wait for a refactor in a future PR, breaking up this method would affect other areas of the code and make the big conditional block even more unwieldy.

**Review by joshdholtz — APPROVED** [review]
This change looks sooooooo good 😊 Appreciate the contribution ❤️ I'll let @janpio merge this one when he thinks its good since he did work on this originally 💪

**Comment by janpio** [comment]
(Thanks @joshdholtz)

**Pull request merged**
